analysis/ic50/__init__.py

```diff
@@ -0,0 +1,90 @@
+from typing import Dict, List, Tuple, Union
+
+import numpy as np
+from scipy.optimize import curve_fit
+from sklearn.metrics import mean_squared_error
+from sklearn.preprocessing import MinMaxScaler
+
+
+def _sigmoid(x: Union[np.array, float], a: float, b: float, c: float, d: float) -> Union[np.array, float]:
+    return ((a - d) / (1.0 + ((x / c) ** b))) + d
+
+
+def _get_starting_parameters(concentrations: np.array, scaled_measurements: np.array) -> List[float]:
+    a = scaled_measurements[0]
+    d = scaled_measurements[-1]
+    b = 1
+    c = concentrations[(np.abs(scaled_measurements - (a + d) / 2)).argmin()]
+    return [a, b, c, d]
+
+
+def _fit_sigmoid(
+        measurements: np.array,
+        scaler: MinMaxScaler, fit_scaler: bool = True,
+) -> Tuple[np.array, np.array, float, MinMaxScaler]:
+    if fit_scaler:
+        scaled_signal = scaler.fit_transform(measurements[:, 1].reshape(-1, 1)).reshape(-1)
+    else:
+        scaled_signal = scaler.transform(measurements[:, 1].reshape(-1, 1)).reshape(-1)
+
+    starting_point = _get_starting_parameters(measurements[:, 0], scaled_signal)
+    try:
+        with np.errstate(invalid='ignore'):
+            fit_param = curve_fit(_sigmoid, measurements[:, 0], np.array(scaled_signal),
+                                  p0=starting_point)
+    except RuntimeError:
+        with np.errstate(invalid='ignore'):
+            fit_param = curve_fit(_sigmoid, measurements[:, 0], np.array(scaled_signal),
+                                  p0=starting_point, maxfev=10000)
```

Comment:
Does `maxfev` will be the only reason of `RuntimeError`? If it doesn't maybe we can somehow check what error we have and don't run `curve_fit` twice

Reply:
AFAIC yes, the only possible reason for a `RuntimeError` is when it fails to converge in a set amount of iterations. I processed several experiments when it's the case for default 1000 iterations, but it converges on 10k (experimental value, which is balanced between convergence and speed). That's the reason I added a "1 more try" on more iterations (usually it's not invoked, so the speed remains OK). And if it fails the second time, then the exception in main file is triggered.